File: runme/mcgen/local/datan/exasrc/e3test.f
      PROGRAM E3TEST
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION R(1000),HIST(100),XPL(101),YPL(101)
      CHARACTER*75 TX,TY,CAPT
C identify program to user
      WRITE(*,*)' Program E3TEST simulates set of measurements'
      WRITE(*,*)' and performs chi-squared test'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number NEPX of experiments (>0)'
      WRITE(*,*)' >'
      READ(*,*)NEXP
      WRITE(*,*)
     +' Enter number N of measurements per experiment (>0)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter mean A for generation'
      WRITE(*,*)' >'
      READ(*,*)A
      WRITE(*,*)' Enter hypothesis A0 for mean'
      WRITE(*,*)' >'
      READ(*,*)A0
      WRITE(*,*)
     +' Enter standard deviation SIGMA for generation (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)
     +' Enter hypothesis SIGMA0 for standard deviation (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA0
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - histogram of chi squared'
      WRITE(*,*)' 2 - histogram of chi-squared probability'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
C initialize histogram
      CAPT(1:33)='n_exp#=*****, n=*****, a=******, '
      CAPT(34:71)='&s@=******, a_0#=******, &s@_0#=******'
      WRITE(CAPT(8:12),'(I5)')NEXP
      WRITE(CAPT(17:21),'(I5)')N
      WRITE(CAPT(26:31),'(F6.2)')A
      WRITE(CAPT(38:43),'(F6.2)')SIGMA
      WRITE(CAPT(51:56),'(F6.2)')A0
      WRITE(CAPT(66:71),'(F6.2)')SIGMA0
      LCAPT=71
      IF(ISWIT.EQ.1) THEN
        H1=DBLE(4*N)
        TX='&c@^2'
        TY='N(&c@^2#)'
        LTX=5
        LTY=9
      ELSE
        H1=1.D0
        TX='P(&c@^2#)'
        TY='N(P)'
        LTX=9
        LTY=4
      END IF
      H0=0.D0
      DELH=0.01D0*H1
      NH=100
      CALL SMHSIN(HIST,H0,DELH,NH)
C loop over all experiments
      DO 30 IEXP=1,NEXP
C simulate sample R
        CALL RNSTNR(R,N)
        DO 10 I=1,N
          R(I)=A+R(I)*SIGMA
10      CONTINUE
C compute chi squared
        CHI2=0.D0
        DO 20 I=1,N
          CHI2=CHI2+((R(I)-A0)/SIGMA0)**2
20      CONTINUE
        IF(ISWIT.EQ.1) THEN
C enter chi squared into histogram
          CALL SMHSFL(HIST,H0,DELH,NH,CHI2,1.D0)
        ELSE
C compute probability and enter it into histogram
          P=1.D0-SCCHI2(CHI2,N)
          CALL SMHSFL(HIST,H0,DELH,NH,P,1.D0)
        END IF
30    CONTINUE
      FACT=(H1-H0)*DBLE(NEXP)/DBLE(NH)
      IF(ISWIT.EQ.1) THEN
C curve of chi-squared distribution
        NPL=NH+1
        DO 40 IPT=1,NPL
          XPL(IPT)=H0+(IPT-1)*DELH
          IF(IPT.EQ. 1.AND. N.EQ.1) XPL(IPT)=1.D-5
          YPL(IPT)=FACT*SDCHI2(XPL(IPT),N)
40      CONTINUE
      ELSE
C curve of constant probability
        XPL(1)=H0
        XPL(2)=H1
        YPL(1)=FACT
        YPL(2)=YPL(1)
        NPL=2
      END IF
C graphical output
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      CALL GRHSCV(XPL,YPL,NPL,HIST,H0,DELH,NH,TX,LTX,TY,LTY,
     +CAPT,LCAPT,NWS)
      END

File: runme/mcgen/local/datan/exasrc/e3test.c
#include <stdio.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static double c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, n, iexp, iswit, nexp, nh, npl, ipt, lcapt, ltx, lty, nws;
    static double delh, fact, hist[100], a, p, r[1000], sigma, a0, h0, h1,
                  sigma0, xpl[101], ypl[101], chi2;
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1, i__2;
    double d__1;
/* identify program to user */
    printf("%s\n", "Program E3TEST simulates set of measurements");
    printf("%s\n\n", "and performs chi-squared test.");
/* ask for input */
    printf("%s\n", "Enter number NEXP of experiments (>0):");
    printf("%s", "> ");
    scanf("%li", &nexp);
    printf("%s\n", "Enter number N of measurements per experiment (>0):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter mean A for generation:");
    printf("%s", "> ");
    scanf("%lf", &a);
    printf("%s\n", "Enter hypothesis A0 for mean:");
    printf("%s", "> ");
    scanf("%lf", &a0);
    printf("%s\n", "Enter standard deviation SIGMA for generation (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Enter hypothesis SIGMA0 for standard deviation (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma0);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - histogram of chi squared");
    printf("%s\n", "2 - histogram of chi-squared probability");
    printf("%s", "> ");
    scanf("%li", &iswit);
/* initialize histogram */
    sprintf(capt, "n_exp#=%5li, n=%5li, a=%6.2lf, "
            "&s@=%6.2lf, a_0#=%6.2lf, &s@_0#=%6.2lf",
            nexp, n, a, sigma, a0, sigma0);
    lcapt = strlen(capt);
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    if (iswit == 1) {
        h1 = (double) (n << 2);
        strncpy(tx, "&c@^2", 5);
        strncpy(ty, "N(&c@^2#)", 9);
        ltx = 5;
        lty = 9;
    } else {
        h1 = 1.;
        strncpy(tx, "P(&c@^2#)", 9);
        strncpy(ty, "N(P)", 4);
        ltx = 9;
        lty = 4;
    }
    h0 = 0.;
    delh = h1 * .01;
    nh = 100;
    smhsin_(hist, &h0, &delh, &nh);
/* loop over all experiments */
    i__1 = nexp;
    for (iexp = 1; iexp <= i__1; ++iexp) {
/* simulate sample R */
        rnstnr_(r, &n);
        i__2 = n;
        for (i = 1; i <= i__2; ++i) {
            r[i - 1] = a + r[i - 1] * sigma;
        }
/* compute chi squared */
        chi2 = 0.;
        i__2 = n;
        for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
            d__1 = (r[i - 1] - a0) / sigma0;
            chi2 += d__1 * d__1;
        }
        if (iswit == 1) {
/* enter chi squared into histogram */
            smhsfl_(hist, &h0, &delh, &nh, &chi2, &c_d1);
        } else {
/* compute probability and enter it into histogram */
            p = 1. - scchi2_(&chi2, &n);
            smhsfl_(hist, &h0, &delh, &nh, &p, &c_d1);
        }
    }
    fact = (h1 - h0) * (double) nexp / (double) nh;
    if (iswit == 1) {
/* curve of chi-squared distribution */
        npl = nh + 1;
        i__1 = npl;
        for (ipt = 1; ipt <= i__1; ++ipt) {
            xpl[ipt - 1] = h0 + (ipt - 1) * delh;
            if (ipt == 1 && n == 1) {
                xpl[ipt - 1] = 1e-5;
            }
            ypl[ipt - 1] = fact * sdchi2_(&xpl[ipt - 1], &n);
        }
    } else {
/* curve of constant probability */
        xpl[0] = h0;
        xpl[1] = h1;
        ypl[0] = fact;
        ypl[1] = ypl[0];
        npl = 2;
    }
/* graphical output */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
    grhscv_(xpl, ypl, &npl, hist, &h0, &delh, &nh, tx, &ltx, ty, &lty, capt,
           &lcapt, &nws);
    return 0;
} /* main */

